pkg/subctl/cmd/root.go

```diff
@@ -85,46 +89,87 @@ func handleNodeLabels() error {
 	_, clientset, err := getClients()
 	panicOnError(err)
 	// List Submariner-labeled nodes
-	selector := labels.SelectorFromSet(labels.Set(map[string]string{"submariner.io/gateway": "true"}))
+	const submarinerGatewayLabel = "submariner.io/gateway"
+	const trueLabel = "true"
+	selector := labels.SelectorFromSet(labels.Set(map[string]string{submarinerGatewayLabel: trueLabel}))
 	labeledNodes, err := clientset.CoreV1().Nodes().List(metav1.ListOptions{LabelSelector: selector.String()})
 	if err != nil {
 		return err
 	}
-	fmt.Printf("There are %d labeled nodes in the cluster\n", len(labeledNodes.Items))
-	for _, node := range labeledNodes.Items {
-		for _, label := range node.GetLabels() {
-			fmt.Printf("Node %s, label %s\n", node.GetName(), label)
+	if len(labeledNodes.Items) > 0 {
+		fmt.Printf("* There are %d labeled nodes in the cluster:\n", len(labeledNodes.Items))
+		for _, node := range labeledNodes.Items {
+			fmt.Printf("  - %s\n", node.GetName())
 		}
-	}
-	if len(labeledNodes.Items) == 0 {
-		// List all nodes and select one
-		allNodes, err := clientset.CoreV1().Nodes().List(metav1.ListOptions{})
+	} else {
+		answer, err := askForGatewayNode(clientset)
 		if err != nil {
 			return err
 		}
-		fmt.Printf("There are %d nodes in the cluster\n", len(allNodes.Items))
-		allNodeNames := []string{}
-		for _, node := range allNodes.Items {
-			allNodeNames = append(allNodeNames, node.GetName())
-		}
-		var qs = []*survey.Question{
-			{
-				Name: "node",
-				Prompt: &survey.Select{
-					Message: "Which node should be used as the gateway?",
-					Options: allNodeNames},
-			},
-		}
-		answers := struct {
-			Node string
-		}{}
+		err = addLabelsToNode(clientset, answer.Node, map[string]string{submarinerGatewayLabel: trueLabel})
+		panicOnError(err)
+
+	}
+	return nil
+}
+
+func askForGatewayNode(clientset kubernetes.Interface) (struct{ Node string }, error) {
+	// List all nodes and select one
+	allNodes, err := clientset.CoreV1().Nodes().List(metav1.ListOptions{})
+	if err != nil {
+		return struct{ Node string }{}, err
+	}
+	fmt.Printf("There are %d nodes in the cluster\n", len(allNodes.Items))
+	allNodeNames := []string{}
+	for _, node := range allNodes.Items {
+		allNodeNames = append(allNodeNames, node.GetName())
+	}
+	var qs = []*survey.Question{
+		{
+			Name: "node",
+			Prompt: &survey.Select{
+				Message: "Which node should be used as the gateway?",
+				Options: allNodeNames},
+		},
+	}
+	answers := struct {
+		Node string
+	}{}
+	err = survey.Ask(qs, &answers)
+	if err != nil {
+		return struct{ Node string }{}, err
+	}
+	return answers, nil
+}
+
+// this function was sourced from:
+// https://github.com/kubernetes/kubernetes/blob/a3ccea9d8743f2ff82e41b6c2af6dc2c41dc7b10/test/utils/density_utils.go#L36
+func addLabelsToNode(c kubernetes.Interface, nodeName string, labels map[string]string) error {
+	const retries = 5
```

Comment:
Nit: 500ms x 5 attempts, 2.5s total of waiting, seems like it might be a bit demanding. Maybe 20 attempts at 500ms? or 1s intervals?

Reply:
yeah, may be it's a good idea to increment that. May be I can even implement the ExponentialBackoff pattern here. 

The problem with labeling nodes, is that kubernetes will update those objects every 10 seconds, with the host status, so it's likely that you could hit an issue, not sure about using patch though, but I saw the examples using retry.